python/mlx/utils.py

```diff
@@ -191,3 +191,46 @@ def tree_unflatten(tree):
         return l
     else:
         return {k: tree_unflatten(v) for k, v in children.items()}
+
+
+def tree_reduce(fn, tree, initializer=None, is_leaf=None):
```

Comment:
I think it is best to simplify this function a bit. The following is exactly the same logic but without keeping an extra `first` flag which is implicitly kept by the accumulator being None or not.

```python
def tree_reduce(fn, tree, initializer=None, is_leaf=None):
    if is_leaf is not None and is_leaf(tree):
        return tree if initializer is None else fn(initializer, tree)

    accumulator = initializer

    if isinstance(tree, (list, tuple)):
        for item in tree:
            accumulator = tree_reduce(fn, item, accumulator, is_leaf)
    elif isinstance(tree, dict):
        for item in tree.values():
            accumulator = tree_reduce(fn, item, accumulator, is_leaf)
    else:
        return tree if accumulator is None else fn(accumulator, tree)

    return accumulator
```

As an aside, the following two-liner also does the same thing.

```python
def tree_reduce(fn, tree, initializer=None, is_leaf=None):
    values = (v for p, v in tree_flatten(tree, is_leaf=is_leaf))
    return functors.reduce(fn, values, initializer)
```

In terms of speed the two-liner is 10%-15% slower for a 128 layer transformer which I think is not too bad but the simplified version of reduce is also quite straightforward so let's keep that.

Finally, I would format the docstring as is in the other functions.